pydomcfg/domzgr/zgr.py

```diff
@@ -0,0 +1,160 @@
+"""
+Base class to generate NEMO v4.0 vertical grids.
+"""
+
+from itertools import product
+
+import xarray as xr
+from xarray import DataArray, Dataset
+
+
+class Zgr:
+    """
+    Base class to generate NEMO vertical grids.
+    """
+
+    def __init__(self, ds_bathy: Dataset, jpk: int):
+        """
+        Initialize class.
+
+        Parameters
+        ----------
+        ds_bathy: Dataset
+             xarray dataset including grid coordinates and bathymetry
+        jpk: int
+             number of computational levels
+        """
+
+        self._bathy = ds_bathy
+        self._jpk = jpk
+
+    # -------------------------------------------------------------------------
+    def _init_ds(self):
+        """
+        Initialise the xarray dataset with empty
+        ``z3{T,W}`` and ``e3{T,W}``
+
+        Returns
+        -------
+        ds: Dataset
+            A copy of the dataset used to initialise the class with new
+            coordinates ``z3{T,W}`` and ``e3{T,W}`` empty dataarrays
+
+        """
+        ds = self._bathy.copy()
+
+        var = ["z3", "e3"]
+        grd = ["T", "W"]
+
+        # Initialise a dataset with empty z3 and e3 dataarrays
+        da = xr.full_like(ds["Bathymetry"], None).expand_dims(z=range(self._jpk))
+        for v, g in product(var, grd):
+            ds[v + g] = da.copy()
+            ds = ds.set_coords(v + g)
+        return ds
+
+    # -------------------------------------------------------------------------
+    def _compute_sigma(self, kindx: DataArray):
+        """
+        Provide the analytical function for sigma-coordinate,
+        a uniform non-dimensional vertical coordinate describing
+        the non-dimensional position of model levels.
+
+        Consider that -1 <= sigma <= 0, with
+
+                 sigma =  0 at the shallower boundary
+                 sigma = -1 at the deeper boundary
+
+        Parameters
+        ----------
+        kindx: DataArray
+            Model levels indexes. Note that
+            *) T-points are at integer values (between 1 and jpk)
+            *) W-points are at integer values - 1/2 (between 0.5 and jpk-0.5)
```

Comment:
No is not, but this is the original docstring of NEMO which is trying to be consistent with the manual. 
In the fortran code, the call to the strecthed-sigma function for W and T level is like this (so consistent with the docstring):

z_gsigw3(ji,jj,jk) = -fssig1( REAL(jk,wp)-0.5_wp, rn_bb )
z_gsigt3(ji,jj,jk) = -fssig1( REAL(jk,wp)       , rn_bb )

This means that the docstring
``
*) T-points are at integer values (between 1 and jpk)
*) W-points are at integer values - 1/2 (between 0.5 and jpk-0.5)
``
refers to the input arguments.

Since our code now does not call the _compute_sigma method separately for W and T grids, I agree it doesnt make any sense that docstring ... I will remove it